R/aggregate_by_vector.R

```diff
@@ -0,0 +1,173 @@
+        if (all(id)) break
+    }
+    ## Last step is always rlm: calculate estimated effects as
+    ## summarised values
+    fit <- rlm(X, expression, ...)
+
+    sampleid <- seq_along(unique(sample))
+    ## This will be needed for NUSE-type of quality control, but will
+    ## need to check for missing data as below.
+    ## se <- unique(summary(fit)$coefficients[sampleid, 'Std. Error'])
+
+    ## Take the sample coefficients ( = summarised expression values)
+    coef  <-  fit$coefficients[sampleid]
+    ## Sort the sample coefficients in the same way as the samplenames
+    ## of expression matrix. Puts NA for the samples without any
+    ## expression value
+    res <- coef[paste0('sample', colnames(x))]
+    names(res) <- colnames(x)
+    res
+}
+
+##' @title Return the Median Polish (Robust Twoway Decomposition) of a matrix
+##'
+##' @description
+##' Fits an additive model (two way decomposition) using Tukey's median
+##' polish procedure using [stats::medpolish()].
+##'
+##' @param x A `matrix` of mode `numeric`.
+##'
+##' @param verbose Default is `FALSE`.
+##'
+##' @param ... Additional arguments passed to [stats::medpolish()].
+##'
+##' @return A `numeric` vector of length identical to `ncol(x)`.
+##'
+##' @author Laurent Gatto
+##'
+##' @family Quantitative feature aggregation
+##'
+##' @export
+##'
+##' @importFrom stats medpolish
+##'
+##' @examples
+##' x <- matrix(rnorm(30), nrow = 3)
+##' medianPolish(x)
+medianPolish <- function(x, verbose = FALSE, ...) {
+    medpol <- stats::medpolish(x, trace.iter = verbose, ...)
+    medpol$overall + medpol$col
+}
+
+##' @title Counts the number of features
+##'
+##' @description
+##' Returns the number of non-NA features in a features by sample
+##' matrix.
+##'
+##' @param x A `matrix` of mode `numeric`.
+##'
+##' @param ... Currently ignored.
+##'
+##' @return A `numeric` vector of length identical to `ncol(x)`.
+##'
+##' @family Quantitative feature aggregation
+##'
+##' @export
+##'
+##' @author Laurent Gatto
+##'
+##' @examples
+##' m <- matrix(c(1, NA, 2, 3, NA, NA, 4, 5, 6),
+##'             nrow = 3)
+##' colCounts(m)
+##' m <- matrix(rnorm(30), nrow = 3)
+##' colCounts(m)
+colCounts <- function(x, ...)
+    colSums(!is.na(x))
+
+
+
+##' @param x A `matrix` of mode `numeric`.
+##'
+##' @param INDEX A `vector` or `factor` of length `nrow(x)`.
+##'
+##' @param FUN A `function` to be applied to the subsets of `x`.
+##'
+##' @param ... Additional arguments passed to `FUN`.
+##'
+##' @return [aggregate_by_vector()] returns a new `matrix` of
+##'     dimensions `length(INDEX)` and `ncol(x), with `dimnames` equal
+##'     to `colnames(x)` and `INDEX`.
+##'
+##' @rdname aggregate
+##'
+##' @export
+aggregate_by_vector <- function(x, INDEX, FUN, ...) {
+    if (!inherits(x, "matrix"))
+        stop("'x' must be a matrix.")
+    if (!identical(length(INDEX), nrow(x)))
+        stop("The length of 'INDEX' has to be identical to 'nrow(x).")
+    INDEX <- factor(INDEX)
```

Comment:
I don't think this line is necessary, from `?tapply`:
> INDEX: a ‘list’ of one or more ‘factor’s, each of same length as
          ‘X’.  The elements are coerced to factors by ‘as.factor’.

Reply:
Ok, thank, will remove it.